README.rst
@@@
dob
@@@

.. image:: https://api.travis-ci.com/hotoffthehamster/dob.svg?branch=proving
  :target: https://travis-ci.com/hotoffthehamster/dob
  :alt: Build Status

.. image:: https://codecov.io/gh/hotoffthehamster/dob/branch/proving/graph/badge.svg
  :target: https://codecov.io/gh/hotoffthehamster/dob
  :alt: Coverage Status

.. image:: https://readthedocs.org/projects/dob/badge/?version=latest
  :target: https://dob.readthedocs.io/en/latest/
  :alt: Documentation Status

.. image:: https://img.shields.io/github/v/release/hotoffthehamster/dob.svg?style=flat
  :target: https://github.com/hotoffthehamster/dob/releases
  :alt: GitHub Release Status

.. image:: https://img.shields.io/pypi/v/dob.svg
  :target: https://pypi.org/project/dob/
  :alt: PyPI Release Status

.. image:: https://img.shields.io/github/license/hotoffthehamster/dob.svg?style=flat
  :target: https://github.com/hotoffthehamster/dob/blob/release/LICENSE
  :alt: License Status

.. |dob| replace:: ``dob``
.. _dob: https://github.com/hotoffthehamster/dob

.. |pip| replace:: ``pip``
.. _pip: https://pip.pypa.io/en/stable/

.. |demo-dob| replace:: *demo dob*
.. _demo-dob: `demo dob`_

.. |install-guide| replace:: installation guide
.. _install-guide: https://dob.readthedocs.io/en/latest/installation.html

.. |update-path| replace:: update ``PATH`` or find where ``pip`` installs ``dob``
.. _update-path: https://dob.readthedocs.io/en/latest/installation.html#update-path

|dob|_ is an interactive, terminal-based time tracking application.

You'll find Vim-like navigation,
robust functionality,
a customizable user experience,
lots of great plugins, and
an elegant, minimalistic interface.

``dob`` is written in `Python <https://www.python.org/>`__
and installs easily with |pip|_:

.. code-block:: sh

    $ pip3 install dob

The best way to learn ``dob`` is to run the demo:

.. code-block:: sh

    $ dob demo

For other setup options, read the |install-guide|_.

- Hint: You might need to |update-path|_.

##########
Quick Peak
##########

dob has lots of CLI commands, and it also includes an interactive terminal application.

Here's a little taste of what the terminal application looks like.::

  ╭────────────────────────────────────────────────────────────────────────────────╮
  │             Thu 16 Apr 2020 ◐ 02:13 PM — 03:52 PM                              │
  ╰────────────────────────────────────────────────────────────────────────────────╯

    duration........... :  1 hour  39 minutes
    start.............. : 2020-04-16 13:13:49
    end................ : 2020-04-16 14:52:50
    activity........... : dobumentation
    category........... : dob Development
    tags............... : #readme

  ┌────────────────────────────────────────────────────────────────────────────────┐
  │I'm adding this Fact to the README. So meta.                                    │
  │                                                                                │
  │                                                                                │
  │                                                                                │
  │                                                                                │
  │                                                                                │
  │                                                                                │
  └────────────────────────────────────────────────────────────────────────────────┘
  Fact ID #22618                              [?]: Help / [C-S]: Save / [C-Q]: Quit

Though in your terminal, you'll see colors and font ornamentation
(bold, italic, and underline).

######
Status
######

.. |beta| replace:: *beta*
.. _beta: https://en.wikipedia.org/wiki/Software_release_life_cycle#Beta

``dob`` is released to |beta|_!

* You can |demo-dob|_ to see how awesome it is!

* You can also start using ``dob``!

  But just be aware that it has not been vetted yet for a larger
  audience — you might still encounter a few issues. But please
  report them, and we'll get them fixed.

  We just don't want you to have a bad first experience.

  Aka, *you've been warned!*

  (Nonetheless, the ``dob`` beta uses a backup file judiciously,
  so if it does crash, you should not lose any unsaved data.)

dob is already feature-rich, and it's used daily (hourly!) by
the authors.

To see what all the fuss is about, you can ``pip3 install dob``
and run ``dob demo`` right now to try it out.

Or, keep reading for an overview of dob, a list of its features,
and then detailed documentation of the interactive editor commands.

OS Support
==========

``dob`` runs on any operating system that supports Python,
including Linux, macOS, and Windows.

Please see the |install-guide|_ for help getting setup.

###########
Quick Start
###########

If you want to dive right in (after running the demo to get acquainted), try::

   $ dob init
   $ dob edit

The ``init`` command creates a config file and the database,
and the ``edit`` command runs the interactive editor.

You can also run ``dob --help`` to see an overview of all commands.

#####
Ethos
#####

``dob`` is a designed for developers, but suitable
for any individual whose comfortable in the terminal.

``dob`` is perfect for the person who asks,
*Why must I use my mouse to manage time tracking?*

``dob`` is even more suited for the person who asks,
*How can I manage time tracking with the fewest possible keystrokes?*

``dob`` is inspired by
`Hamster <https://github.com/projecthamster/>`__,
a beloved but aged time tracking application for
`GNOME <https://en.wikipedia.org/wiki/GNOME>`__.

``dob`` is emulative of `Vim <https://www.vim.org/>`__,
an editor that enables users to concentrate on their
work while the tool itself slips into the background.

Try ``dob`` today -- you might like it!

- `Demo dob`_ to learn it.

- | Then, `start from scratch`__,
  | or `load a legacy database`__.

__ https://dob.readthedocs.io/en/latest/usage.html#start-fresh

__ https://dob.readthedocs.io/en/latest/usage.html#upgrade-hamster

If you like |dob|_, hopefully you'll
`help us make it better
<https://dob.readthedocs.io/en/latest/contributing.html>`_!

########
Features
########

- Minimalist interactive console app fits elegantly into your terminal-based workflow.

- Robust configuration -- and manageable via CLI if you want to avoid the config file.

- Fully customizable -- change which keys map to which commands.

- Define your own user-specific paste commands -- assign your own activities and tags.

- Undo and redo edits as you work -- don't worry if you make a mistake.

- Auto-complete makes it easy to set activities and tags -- and shows useful stats, too.

- Edit text in your favorite ``$EDITOR`` -- and set a file extension for highlighting.

- Use all the Unicode you want -- spice up your notes!

- Personalize the interface colors -- you can even style your own activity and tag names!

- Natural syntax lets you go "offline" and dob to a text file, that you can import later.

- Migrates legacy Hamster databases -- and repairs integrity issues, too.

- Supports unrestricted plugin access -- anyone can add their own commands and config.

- Includes command shortcuts -- common command sequences mapped to a single keypress.

- Runs on the latest Python (3.6, 3.7, and 3.8).

- Broad desktop OS support -- runs on Linux, macOS, and Windows.

- Suitably tested (though coverage could be better).

- Well documented -- get help at runtime, on the command line, or online.

- Simple, smooth code base -- follows best practices, and uses top libraries.

- Low bar of entry to get raw access to data --
  defaults to `Sqlite3 <https://www.sqlite.org/index.html>`_ file.

- Constant dogfooding *(-- I dob daily!)*

- Free and open source -- hack away!

####
Demo
####

Demo ``dob``
============

Run the ``demo`` command to load an interactive tutorial:

.. code-block:: sh

    $ dob demo

And then follow the walk-through.

###############
Getting started
###############

Read `basic usage`__ to learn how to create an empty data store,
or how to import an existing database.

__ https://dob.readthedocs.io/en/latest/usage.html

############
Learning dob
############

You'll find lots of documentation online, including:

- `Configure dob`__

- `Choose your $EDITOR`__

- `Run the Interactive Editor`__

- `Exit, Save, and Undo/Redo`__

- `Navigate Facts Quickly`__

- `Edit Facts and Metadata`__

- `Copy and Paste Metadata`__

- `Useful Command Combinations`__

- `Nudge Start and End Time`__

__ https://dob.readthedocs.io/en/latest/guide-config.html
__ https://dob.readthedocs.io/en/latest/guide-editor-env.html
__ https://dob.readthedocs.io/en/latest/guide-intro-cli-and-editor.html
__ https://dob.readthedocs.io/en/latest/guide-exit-save-undoredo.html
__ https://dob.readthedocs.io/en/latest/guide-jumping-around.html
__ https://dob.readthedocs.io/en/latest/guide-editing-facts.html
__ https://dob.readthedocs.io/en/latest/guide-copy-paste.html
__ https://dob.readthedocs.io/en/latest/guide-combinations.html
__ https://dob.readthedocs.io/en/latest/guide-nudging-time.html

#######
Plugins
#######

Plugins make it easy for everyday dobbers to write their own
features and to share them with the broader community.

The core development team has created the following plugins
for non-essential and distribution-specific features.

*Please note: these plugins are not yet published!*
(We just want to get you excited in the meantime.)

- The ``dob-plugin-export-commit``
  plugin exports and commits changes to your dob database on every save.

- The ``dob-plugin-stale-fact-goader``
  plugin displays a GNOME-style popup notification after some amount of
  inactivity to badger you to start a new Fact.

- The ``dob-plugin-git-hip``
  plugin tags the active Fact based on the Git branch name
  pulled from the current directory.

Have a great idea for ``dob``? Implement it as a plugin!

- Plugins have complete access to dob.

  Plugins can add their own commands.

  Plugins can add their own configuration.

  Plugins can run on startup, on exit, or whenever the user saves.

Plugins help keep ``dob`` lean, and they let you turn
an idea into a feature quickly and easily!

#######
Thanks!
#######

We hope you enjoy using dob as much as we have had making it!

And if you like it, please tell your friends and colleagues about it.
Tell the whole world!!

Keep on dobbin'!



HISTORY.rst
#######
History
#######

.. |dob| replace:: ``dob``
.. _dob: https://github.com/hotoffthehamster/dob

.. |dob-viewer| replace:: ``dob-viewer``
.. _dob-viewer: https://github.com/hotoffthehamster/dob-viewer

.. |dob-prompt| replace:: ``dob-prompt``
.. _dob-prompt: https://github.com/hotoffthehamster/dob-prompt

.. |dob-bright| replace:: ``dob-bright``
.. _dob-bright: https://github.com/hotoffthehamster/dob-bright

.. |nark| replace:: ``nark``
.. _nark: https://github.com/hotoffthehamster/nark

.. |hamster-cli| replace:: ``hamster-cli``
.. _hamster-cli: https://github.com/projecthamster/hamster-cli

.. |ohmyrepos| replace:: OhMyRepos
.. _ohmyrepos: https://github.com/landonb/ohmyrepos

.. :changelog:

3.0.13 (2020-06-18)
===================

- Highlights: Much improved searching and reporting.

  - A brief overview of what's documented in greater detail below:

    - Add grouping (by Activity, Category, Tags, and/or start date).

      - For instance, group by Activity, Category, and Day to see
        how much time was spent on each Activity\@Category each day.

    - Add multi-column sorting.

      - For instance, group by Activity, Category, and Day, and sort
        by day and usage to see results ordered by which tasks had the
        most time spent on them recently.

    - Add search on Fact description.

      - For instance, find all Facts whose description contains one
        or more search terms.

    - Add Tag frequency distributions.

      - For instance, to see the number of times each Tag was used in
        each result group.

    - Add JSON output format.

      - For instance, to prepare data to transmit elsewhere, such as
        <third-party timesheet server>.

    - Support human-friendly relative dates (like '1 day ago').

      - E.g., ``dob find --since yesterday``.

    - Wire since/until options to Activity, Category, and Tag searches
      (via ``list`` and ``usage`` commands).

      - E.g., ``dob list activities --since 'last week'``.

    - New 'Journal' search results format, exemplified by the new
      ``dob report`` command.

      - Includes sparklines and Tag frequencies, e.g.::

          $ dob report
          Fri Jun 12  8.3 hours  ████▌    Development@Tally Bark  #dob(5)
                      2.5 hours  █▍       Cooking & Eating@Personal  #salad

- Details: Add more powerful search and report options.

  - Rename a few existing options and a handful of single-character
    options, and add a number of new options for the search commands
    (for the existing ``search``, ``list``, ``usage``, and ``export``
    commands, and for the new ``report`` command).

  - Bring parity to the ``list``, ``usage``, ``search`` and ``export``
    commands, such that they all support as many of the same options as
    reasonable.

  - Add new report grouping options:

    - New ``-A/--group-activity``, to group results by Activity.

    - New ``-C/--group-category``, to group results by Category.

    - New ``-T/--group-tags``, to group results by Tag names.

    - New ``-Y/--group-days``, to group by Fact start date (YYYYMMDD).

  - Allow multiple columns to be specified with ``-S/--sort/--order``.

  - Remove ``-D/--desc`` and ``-A/--asc`` options and rely instead on
    ``-D/--direction/--dir`` option, which can now be specified multiple
    times, to support specifying a separate sort order for each of the
    now-multiple-allowed ``-S/--sort/--order`` options.

  - Rename report option ``-g/--category`` to ``-c/--category`` (which
    restricts results to matching Category), so that the option name
    pneumonic is obvious.

  - Add new ``-t/--tag`` search match option, to complement existing
    ``-a/--activity`` and ``-c/--category`` search match options.

  - Add search term support, for matching against Fact description.

    - E.g., ``dob find foo`` will find all Facts with 'foo' in their
      description.

  - Add single-character options for ``--since`` and ``--until`` options:
    ``-s`` and ``-u``.

  - Disable deprecated report options, ``--hidden`` and ``--deleted``.

  - Rename option ``-u/--usage`` to ``-U/--show-usage``.

  - Rename option ``-s/--span/--no-span`` to ``-N/--hide-duration``.

  - New option, ``-P/--hide-description``, to omit the Fact description.

  - New option, ``-l/--column``, allows user to specify exactly what
    details to include in the report (otherwise a reasonable set of
    default columns is reported).

    - E.g., the command,

      ``dob list facts --group-tags -l 'tag' -l 'group_count'``

      will show two columns, 'Tag' and 'Uses', that indicate
      each Tag name used on the grouped Facts, as well as how
      many Facts it's used on, and excludes all other columns.

  - Add a new sparkline output value, to visually represent the 'duration'.

    - And add options to control the new sparkline output value:
      ``--spark-width``, ``--spark-total``, and ``--spark-secs``.

  - Rename option ``-r/--rule`` to ``-R/--factoid-rule``, for readability.

  - Replace poorly implemented ``-t/--truncate`` option with ``-W/--width``
    option, to restrict report table or Factoid line to specific width.

    - Also change Factoid report to default to compact view (no blank
      lines) when width is used.

      - E.g., ``dob find --since 'last week' --factoid --width 110``.

  - Rename poorly-named ``-w/--doc`` output format option to ``--factoid``
    (or use ``-f factoid``).

  - Add JSON output format option (specify with ``-f json``, ``--json``,
    or ``-J``).

  - Split table and markup formats from ``--format`` to a new option,
    ``--table-type/--type``.

    - Now the ``-f/--format`` option only includes the higher-level
      formats, which are also each mapped to their own option names,
      e.g., ``--journal`` is equivalent to ``--format journal``.

      This list of higher-level formats is:

      ``--journal``, ``--factoid``, ``--ical``, ``--csv``, ``--json``,

dob/complete.py
# This file exists within 'dob':
#
#   https://github.com/hotoffthehamster/dob
#
# Copyright © 2018-2020 Landon Bouma. All rights reserved.
#
# 'dob' is free software: you can redistribute it and/or modify it under the terms
# of the GNU General Public License  as  published by the Free Software Foundation,
# either version 3  of the License,  or  (at your option)  any   later    version.
#
# 'dob' is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
# without even the implied warranty of MERCHANTABILITY  or  FITNESS FOR A PARTICULAR
# PURPOSE.  See  the  GNU General Public License  for  more details.
#
# You can find the GNU General Public License reprinted in the file titled 'LICENSE',
# or visit <http://www.gnu.org/licenses/>.

import os
import random
import re
import sys

from gettext import gettext as _

import click_hotoffthehamster as click
from click_hotoffthehamster.parser import split_arg_string
from nark.helpers.parsing import (
    ParserMissingActivityException,
    ParserMissingDatetimeOneException,
    ParserMissingDatetimeTwoException
)
from nark.items.fact import Fact

__all__ = ('tab_complete', )


TIME_HINT_MAP = {
    # SYNC_ME: RE_TIME_HINT, TIME_HINT_MAP, and @generate_add_fact_command's.
    'on': 'verify_none',
    'now': 'verify_none',  # 'on' alias
    'from': 'verify_both',
    'between': 'verify_both',  # 'from' alias
    'at': 'verify_start',
    'to': 'verify_end',
    'until': 'verify_end',  # 'to' alias
    'then': 'verify_then',
    'still': 'verify_still',
    'after': 'verify_none',
    'next': 'verify_none',  # 'after' alias
}


# FIXME/2018-05-17 18:04: On `dob import <tab>`, showing commands, not files! why??

def tab_complete(controller):
    def _complete_cmd(controller):
        args, incomplete = _do_complete()
        if len(args) < 1:
            return

        try:
            time_hint = TIME_HINT_MAP[args[0]]
        except KeyError:
            # Not an add_fact command! Fall-back to Click's complete.
            return

        for item in _get_choices(args[1:], incomplete, time_hint):
            click.echo(item)

    # NOTE: (lb): This fcn. was copied from click/_bashcomplete.py::do_complete().
    #       Then it was modified.
    def _do_complete():
        try:
            cwords = split_arg_string(os.environ['COMP_WORDS'])
        except KeyError:
            controller.client_logger.error(_(
                'Expected environ "COMP_WORDS" not set.'
            ))
            sys.exit(1)
        try:
            cword = int(os.environ['COMP_CWORD'])
        except ValueError:
            # This should only happen if your testing the complete command.
            assert os.environ['COMP_CWORD'] == ''
            cword = 0
        args = cwords[1:cword]
        try:
            incomplete = cwords[cword]
        except IndexError:
            incomplete = ''
        return (args, incomplete)

    def _get_choices(args, incomplete, time_hint):
        choices = []
        try:
            # Meh. Probably don't need FactDressed.
            fact, _err = Fact.create_from_factoid(args, time_hint=time_hint)
        except ParserMissingDatetimeOneException as err:
            choices = _choices_datetimes(controller, incomplete, time_hint, err)
        except ParserMissingDatetimeTwoException as err:
            choices = _choices_datetimes(controller, incomplete, time_hint, err)
        except ParserMissingActivityException:
            # See if we should suggest tags, activities, or activity-categories.
            if (not incomplete) or (not re.match(r'''^['"]?[#@]''', incomplete)):
                choices = choices_activities(
                    controller, incomplete, whitespace_ok=False,
                )
            else:
                choices = choices_tags(
                    controller, incomplete, whitespace_ok=False,
                )
        else:
            choices = [
                'Now_You_Write_A_Description',
                'Where-do-hamsters-go-after-graduation?__Hamsterdam!',
            ]

        return choices

    # The shim to the inline main(), above.
    _complete_cmd(controller)


def _choices_datetimes(controller, incomplete, time_hint, parser_err):
    """Suggest times."""
    now = controller.now
    # Show friendly usage reminders.
    # - For 'verify_start' (dob-at), show now and very recent times.
    # - For 'verify_end' (dob-to-/-until), show now (and very recent).
    # - For 'verify_both', with show less freshly recent for start,
    #   and show more recent times for second, end time.
    # NOTE: We underscore because Bash complete splits words,
    #       and thankfully the friendly `dateparser` understand this
    #       (for the most part; e.g., 'three years ago' works, but
    #       not 'three_years_ago', though '3_years_ago' works, go figure).
    friendly_at_hints = [
        '10_minutes_ago',
        '5_seconds_ago',
        'at_1_pm',
        'an_hour_ago',
        'noon',
        'midnight',
    ]
    friendly_from_hints = [
        # (lb): The shell splits completes on whitespace, so use underscores.
        # FIXME: (lb): Update parser to convert underscores back to whitespace
        #        before passing to human friendly time parser.
        'yesterday_at_3_PM',
        '1_week_ago',
        '3_years_ago',
        'Monday_at_16:44',
        'at_noon'
        '1_week_ago_at_midnight',
        # Hahaha, futuristic!
        # (lb): This is an invalid option. But it's funny!
        'tomorrow',
    ]

    # Show one friendly date example, randomly selected.
    # (lb): Using random.shuffle, rather than random.choice. Can't remember why.
    random.shuffle(friendly_at_hints)
    random.shuffle(friendly_from_hints)
    if time_hint == 'verify_start':
        friendly_hint = friendly_at_hints[:2]
    elif time_hint == 'verify_end':
        friendly_hint = ['now']
    elif time_hint == 'verify_both':
        if isinstance(parser_err, ParserMissingDatetimeOneException):
            friendly_hint = friendly_from_hints[:2]
        else:
            assert isinstance(parser_err, ParserMissingDatetimeTwoException)
            friendly_hint = friendly_at_hints[:2]
            friendly_hint.append('now')

    choices = [
        now.strftime('%Y-%m-%d'),
        now.strftime('%H:%M'),
        # Bash will split the completion word in a space,
        # so use \S+ form of YYYY-MM-DD and %H:M.
        # MEH: dateparser recognizes almost any delimiter
        # between the YYYY-MM-DD and the %H:%M, except
        # '-' and '+' indicate time zone. 'T' seems fine.
        now.strftime('%Y-%m-%dT%H:%M'),
    ]
    choices += friendly_hint
    if incomplete:
        choices = [
            ch for ch in choices if not incomplete or ch.startswith(incomplete)
        ]
    return choices


def choices_tags(controller, incomplete='', whitespace_ok=False):
    """Suggest tags."""
    # Grab the last 20 or so tags used (by Facts, chronologically),
    # and also the most used (top ten) 10 or so tags used by all Facts.
    #
    # FIXME: Make these limits settable (via config?).
    # FIXME: Can we cycle through the various sort options?
    #        E.g., if user TABs, show list of previous used tags.
    #          And if user TABs a second time, show list of most
    #          use tags. TAB again, another sort option, etc.
    # MAGIC_NUMBERS: 21 and 13, eh. Arbitrary limits.
    tags_counts = controller.tags.get_all_by_usage(sort_cols=('start',), limit=21)
    tags_counts += controller.tags.get_all_by_usage(sort_cols=('usage',), limit=13)

    choices = [
        '@{}'.format(tag.name) for tag, _uses, _span
        in tags_counts if not incomplete or tag.name.startswith(incomplete[1:])
    ]
    # MEH: We cull, even though we set limit above, so total count might be even smaller.
    #   (We could solve at the SQL query level, but who wants to go to the trouble?)
    if not whitespace_ok:
        choices = [tag for tag in choices if ' ' not in tag]
    return choices


def choices_activities(
    controller,
    incomplete='',
    whitespace_ok=False,
):
    """Suggest activities."""

    acty = controller.activities.get_all_by_usage(sort_cols=('start'))

    choices = [
        '{}@{}'.format(
            act.name, act.category.name if act.category else ''
        ) for act, count in acty
    ]

    # (lb): Bash complete doesn't handle spaces well, so ignore
    # those activities@categories with and spaces.
    if not whitespace_ok:
        choices = [atc for atc in choices if ' ' not in atc]
    if incomplete:
        choices = [atc for atc in choices if atc.startswith(incomplete)]

    # FIXME/2018-05-15 23:35: Caller should specify max, or just deal with
    # all results. E.g., tab-complete only wants so many. E.g., --ask only
    # wants as many can fit screen (or all, if I implement pagination). (lb)
    max_choices = 50
    if len(choices) > max_choices:
        choices = choices[:50]

    return choices

